### PR TITLE
Accepts wild card (*) in `indexes` field for creating Keys and tenant tokens (end of the word only)

### PR DESCRIPTION
**Pull request by akhildevelops**
# Pull Request

## Related issue
Fixes partially:  [#2788 Use globing wildcard (*) when creating API keys](https://github.com/meilisearch/meilisearch/issues/2788)

## What does this PR do?
Mellisearch's '/keys' endpoint can accept index wild card patterns.
Example:
```
{
    "indexes": ["products_*"],
    "actions": ["documents.*"],
    "expiresAt": "2042-04-02T00:42:42Z"
  }
```

The generated API Key can perform all *documents* related actions on any index that matches to `products_*`.

This PR also includes an integration test case.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Review by curquiza — COMMENTED** [review]
Thanks a lot @Enforcer007 for your PR!

From my perspective, so only the user interface, it works, <s>but only for `indexname_*` pattern so the `*` valid at the end</s> -> this is the only pattern we want to accept as external contribution, so it's ok.

So this issue only fix partially #2869, which is already great! <s>@Enforcer007 I cannot remove the issue from the PR in the `Developement` section on the right, can you do it please?</s> Looks like it's not anymore, np!

Need the developers review!

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
@Enforcer007 can you fix the Clippy error before the core team reviews your PR? 😇

File: meilisearch-auth/src/store.rs
Comment (Kerollmops) [inline meilisearch-auth/src/store.rs:217]: I am not sure but using equal here is probably not idiomatic I would prefer that we introduce a `match` method or something, don't you think?

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Overall, I find this PR very good. Thank you for contributing!

However, there is a lot of implementation of `Eq` and `PartialEq` traits that I would change into a single `match` method, maybe.

Also, I would like that you add more documentation to the new types you introduced.

And could you explain why you touched the heed-store codecs by removing the index field from it? Where do you store the pattern now?

File: meilisearch-auth/src/store.rs
Comment (akhildevelops) [reply meilisearch-auth/src/store.rs:218]: x is a struct with StarOr<IndexType> type which is a representation of index name / pattern. It would be appropriate in comparing the x with string.

File: meilisearch-auth/src/store.rs
Comment (akhildevelops) [reply meilisearch-auth/src/store.rs]: Considered

File: meilisearch-types/src/index_uid.rs
Comment (akhildevelops) [reply meilisearch-types/src/index_uid.rs]: Considered

File: meilisearch-types/src/index_uid.rs
Comment (akhildevelops) [reply meilisearch-types/src/index_uid.rs]: Considered

File: meilisearch-auth/src/store.rs
Comment (akhildevelops) [reply meilisearch-auth/src/store.rs:277]: If we use `(key_id,action,index)` model, we can't authenticate the api key from `action_keyid_index_expiration` database. when the index is a pattern. In current implementation, the authentication fails, if an index matches to index pattern.

**Comment by akhildevelops** [comment]
Thanks @Kerollmops for the feedback. I've responded to your comments. Pls review the updated PR. 

@curquiza fixed clippy warnings / errors.

**Comment by akhildevelops** [comment]
@curquiza can u pls rerun workflow

**Comment by akhildevelops** [comment]
bors try

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
@akhildevelops looks like there are issues also about the tests 😕

**Comment by akhildevelops** [comment]
@curquiza  Now please run, fixed them.

**Comment by akhildevelops** [comment]
Instead of manually approving to run workflows everytime.

I would suggest one approval per PR and thereafter it should run automatically for all the updates.

File: meilisearch-auth/src/store.rs
Comment (curquiza) [inline meilisearch-auth/src/store.rs:7]: If you don't use it, can you delete it please?

**Review by curquiza — DISMISSED** [review]

File: meilisearch-auth/src/store.rs
Comment (akhildevelops) [reply meilisearch-auth/src/store.rs]: Done

**Comment by curquiza** [comment]
No probelm @akhildevelops! We will review it once it's fixed 😄

**Comment by curquiza** [comment]
bors try

**Review by curquiza — APPROVED** [review]
This is still working as expected on my side, I made some test

@Kerollmops I let you review before merging if it's ok for you

thank you again @akhildevelops 😊

**Comment by curquiza** [comment]
Yes your review is planned @akhildevelops
Sorry we have a lot to do this month in additionally to Hacktoberfest

I put @ManyTheFish instead of @Kerollmops to ensure you will be reviewed

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
Hello @akhildevelops, thank you for your contribution,
I requested some changes, mainly in the tests to ensure that no corner cases are forgotten.
I let you make the changes, ping me when you are ready 👍

**Comment by akhildevelops** [comment]
@ManyTheFish This PR currently supports `documents.*` actions. Validating for other actions.

**Comment by ManyTheFish** [comment]
@akhildevelops, sorry I'm probably missing something, I don't understand your last comment.

**Comment by akhildevelops** [comment]
My changes currently support doing `docuemnts.*` related actions. https://docs.meilisearch.com/reference/api/documents.html

Tests are running successfully, inserting documents to indexes that match to the index pattern. 

But tests fail to create indexes like products_1, products_2 that match to products_* pattern. I'm currently validating for other APIs `tasks.*, settings.*, stats.*`. I'll ping back when everything is fine.

**Comment by curquiza** [comment]
@akhildevelops we wil not have the time to merge your PR by the end of 31st so I put the `hackotberfest-accepted` label on your PR for the work you made! thanks for your PR!

**Comment by curquiza** [comment]
@akhildevelops we still want to use your work, can you put your commits again and re-open the PR, otherwise I cannot let the label `hacktoberfest-accepted`

**Comment by akhildevelops** [comment]
Thanks @curquiza, Somehow it got closed, reopening it.

**Comment by akhildevelops** [comment]
Resolving merge conflicts. Will give a shoutout after everything is streamlined.

**Comment by akhildevelops** [comment]
@curquiza Can you pls rerun the workflow. I've tested locally and everything is working fine.

@ManyTheFish I've added the test cases. Pls review

- [lazy_create_index_from_pattern](https://github.com/akhildevelops/meilisearch/blob/e2097b3ee571705f5d099598e0158bbf2e491fcc/meilisearch-http/tests/auth/authorization.rs#L832-L929)
- [error_creating_index_without_index](https://github.com/akhildevelops/meilisearch/blob/e2097b3ee571705f5d099598e0158bbf2e491fcc/meilisearch-http/tests/auth/authorization.rs#L932-L1005)
- [access_authorized_index_patterns](https://github.com/akhildevelops/meilisearch/blob/e2097b3ee571705f5d099598e0158bbf2e491fcc/meilisearch-http/tests/auth/authorization.rs#L420-L482)
- [raise_error_non_authorized_index_patterns](https://github.com/akhildevelops/meilisearch/blob/e2097b3ee571705f5d099598e0158bbf2e491fcc/meilisearch-http/tests/auth/authorization.rs#L485-L566)
- [pattern_indexes](https://github.com/akhildevelops/meilisearch/blob/e2097b3ee571705f5d099598e0158bbf2e491fcc/meilisearch-http/tests/auth/authorization.rs#L569-L601)

**Comment by curquiza** [comment]
bors try

**Comment by curquiza** [comment]
bors try

File: meilisearch-http/tests/auth/authorization.rs
Comment (ManyTheFish) [reply meilisearch-http/tests/auth/authorization.rs:604]: Up this comment, `list_authorized_tasks_restricted_index` is untested with index patterns

**Review by ManyTheFish — CHANGES_REQUESTED** [review]
Hello @akhildevelops,
I dug into your code and in your tests, there are several minor changes to do.
I saw that you've created a new error code, this kind of thing is decided on the product-team side and I don't manage to find any approval from them about creating this error code. 🤔
If I'm right, I'd prefer to remove this change.

I let you make the fixes! 👍 

> since this is quite a big PR and you made a lot of changes we wanted to thank you for your time and accept your contribution for the hacktoberfest even though we're probably not going to merge it before the end of the event.
That's why I gave you the hacktoberfest-accepted tag.
And if you would like to receive some swag from Meilisearch, please complete [this form](https://meilisearch.typeform.com/to/idt5qugY).

**Pull request closed without merging**
